expat/tests/common.c

```diff
@@ -191,18 +192,18 @@ enum XML_Status
 _XML_Parse_SINGLE_BYTES(XML_Parser parser, const char *s, int len,
                         int isFinal) {
   const int chunksize = g_chunkSize;
-  int offset = 0;
   if (chunksize > 0) {
     // parse in chunks of `chunksize` bytes as long as possible
-    for (; offset + chunksize < len; offset += chunksize) {
-      enum XML_Status res = XML_Parse(parser, s + offset, chunksize, XML_FALSE);
+    for (; chunksize <= len; len -= chunksize, s += chunksize) {
+      enum XML_Status res = XML_Parse(parser, s, chunksize, XML_FALSE);
       if (res != XML_STATUS_OK) {
         return res;
       }
+      assert(s != NULL); // or "s += chunksize" would be undefined behavior
```

Comment:
I think an `assert(s != NULL || len == 0)` or `assert(! (s==NULL && len > 0))` at the start of the function or above the loop might be clearer, but this is also fine.

Reply:
@Snild-Sony I went even further with the 180° turn approach from https://github.com/libexpat/libexpat/pull/787#discussion_r1388499466 now. Pushing in a second…